File: src/jvmMain/java/core/ProcessStreamGlobber.java
package core;

/* Simone 08/07/2014 12:47 */

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;

public class ProcessStreamGlobber {
    final static Logger log = LoggerFactory.getLogger(ProcessStreamGlobber.class);
    private Process process;
    private String name;

    public ProcessStreamGlobber(Process process) {
        this.process = process;
    }

    public void startGlobber() {
        startStreamGobbler(process.getErrorStream(), "ERR");
        startStreamGobbler(process.getInputStream(), "OUT");
    }

    private void startStreamGobbler(InputStream errorStream, String streamName) {
        StreamGobbler sg = new StreamGobbler(errorStream, streamName);
        if (name != null)
            sg.setName(name + "-" + streamName);
        sg.start();
    }

    public void setName(String name) {
        this.name = name;
    }

    class StreamGobbler extends Thread {
        InputStream is;
        String type;

        StreamGobbler(InputStream is, String type) {
            this.is = is;
            this.type = type;
        }

        public void run() {
            try {
                InputStreamReader isr = new InputStreamReader(is);
                BufferedReader br = new BufferedReader(isr);
                String line;
                while ((line = br.readLine()) != null)
                    log.info(type + "> " + line);

            } catch (IOException ioe) {
                ioe.printStackTrace();
            }
        }
    }
}

File: src/jvmMain/kotlin/core/ProcessStreamGlobber.kt
package core

import org.slf4j.LoggerFactory
import java.io.InputStream

/* Simone 08/07/2014 12:47 */

class ProcessStreamGlobber(private val process: Process, val processName: String = "") {
    companion object {
        private val log = LoggerFactory.getLogger(ProcessStreamGlobber::class.java)
        fun handleIS(inputStream: InputStream, type: String) {
            inputStream
                .bufferedReader()
                .lines()
                .filter { it != null }
                .forEach { log.info("$type> $it") }
        }
    }

    fun startGlobber() {
        startStreamGobbler(process.errorStream, "ERR")
        startStreamGobbler(process.inputStream, "OUT")
    }

    private fun startStreamGobbler(errorStream: InputStream, streamName: String) {
        val sg = StreamGobbler(errorStream, streamName)
        if (processName.isNotEmpty()) sg.name = "$processName-$streamName"
        sg.start()
    }

    internal inner class StreamGobbler(val input: InputStream, val type: String) :
        Thread() {
        override fun run() {
            handleIS(input, type)
        }

        init {
            isDaemon = true
        }
    }


}